Closes #2573: Upgrade Drupal core to 10.1.x.

joeparsons opened the pull request:
## Description

- Upgrades Drupal core to 10.1.x.
- Adds new permissions for deleting files to content admin and content editor roles
- Addresses some code deprecations, suppresses some phpstan false-positives
- Updates `.eslintrc.json` config to more closely match 10.1.x version included in Drupal core
- Applies various eslint fixes to `.js` and `.yml` files
- ~Updates lando tooling for eslint to work the same way it the eslint check in Probo does~
- Updates Probo config to work with Drush 12
- Updates linkit module to a Drupal 10.1.x compatible version
- Removes core patches that are no longer needed, updates patches that no longer applied cleanly to 10.1.x
- Adds patch for honeypot module issue affected by change in core's DB transaction isloation level default settings
- Removes no-longer-neeed claro block placement configs from install profile
- Removes `Custom Block Library` view from `/admin/content` since core now provides an equivalent view
- Updates node form style overrides in `az_core` to work with updated version of Claro admin theme

## Related issues
Closes #2573 

## How to test
<!--- Please describe in detail how reviewers can test your changes -->
<!--- Include details of your testing environment and the tests you ran -->
- composer install
- Basic site testing
- Test deleting files.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [x] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [x] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

tadean commented:
```
Call to deprecated method isEmpty() of class                                   
         Drupal\Component\Diff\Diff:                                                    
         in drupal:10.1.0 and is removed from drupal:11.0.0. There is no                
         replacement.
```
Looks like Drupal core is[ planning on removing](https://www.drupal.org/node/3337942) part of their Diff Engine in favor of switching some of the functionality to `sebastianbergmann/diff`.                                                                   

Our config provider uses this internally to check for changes in override config.

joeparsons commented:
> ```
> Call to deprecated method isEmpty() of class                                   
>          Drupal\Component\Diff\Diff:                                                    
>          in drupal:10.1.0 and is removed from drupal:11.0.0. There is no                
>          replacement.
> ```
> 
> Looks like Drupal core is[ planning on removing](https://www.drupal.org/node/3337942) part of their Diff Engine in favor of switching some of the functionality to `sebastianbergmann/diff`.
> 
> Our config provider uses this internally to check for changes in override config.

Thanks!  Yeah, I ended up basically copying that deprecated method as a [new protected method](https://github.com/az-digital/az_quickstart/pull/2587/files#diff-a4ea43d0d65fc95d5831f77ebb88707603e11d521671e100b2064ae5459e23acR249-R264) in our config provider class.

trackleft commented:
We should update our roles to add these two
Delete any file (content administrator)
Delete own files (content administrator)

joeparsons commented:
> We should update our roles to add these two Delete any file (content administrator) Delete own files (content administrator)
> 
> See https://www.drupal.org/node/3324270

Thanks.  Added these permissions:
- Added `delete any file` to Content Admin
- Added `delete own files` to Content Editor

joeparsons commented:
Added some `// phpstan-ignore-next-line` comments to suppress false-positive phpstan deprecation errors related to Drupal core's `TranslatableRevisionableStorageInterface`.  The issue that is generating these false-positive deprecations has been addressed in phpstan-drupal 1.2.0 and Drupal 10.2.x so we should be able to remove these after we upgrade to 10.2.x.

https://www.drupal.org/project/drupal/issues/3383215
https://github.com/mglaman/phpstan-drupal/releases/tag/1.2.0

joeparsons commented:
This scaffolding repo PR should be merged at the same time as this one: https://github.com/az-digital/az-quickstart-scaffolding/pull/117

natebyerley commented:
Site is behaving well overall.

File delete works great.  I embedded a .pdf on a page, then deleted it, and the entity on the page disappeared from front end view (was still present/visible) on backend.

One issue I noticed on probo build was a strange formatting for page editing:

![Screenshot 2023-09-13 at 12 57 39 PM](https://github.com/az-digital/az_quickstart/assets/97116860/f0f31f6b-f8f7-4dbb-a90e-db5f1bf97b50)

joeparsons commented:
> Site is behaving well overall.
> 
> File delete works great.
> 
> I uploaded a .pdf, embedded it on a page, then deleted it from Content > File, and the entity on the page disappeared from front end view (was still present/visible) on backend as basically an entity with a invalid entity reference - this seems like expected behavior to me.
> 
> One issue I noticed on probo build was a strange formatting for page editing:

Thanks @natebyerley!  I'll look into the page editing UI issue.

trackleft commented:
For the release notes

Changes to site-owner-managed files
[The root .htaccess file now unsets the X-Content-Type-Options header before setting it again](https://www.drupal.org/node/3334163). This prevents duplicate headers in some configurations of Apache. Site owners should update their .htaccess files with this change to avoid duplicate headers.
[The root .htaccess file now caches all files for one year instead of two weeks.](https://www.drupal.org/project/drupal/issues/3311406) This brings the value in line with industry standards.
[Drupal adds 'Samesite: Lax' as a session cookie attribute by default](https://www.drupal.org/node/3275352). This is configurable in default.services.yml and site owners should update their copy of the file to include the section.
Sites using nginx and php-fpm may need to update their nginx.conf for changes to [CSS and JavaScript aggregation](https://www.drupal.org/node/2888767#nginx-php-fpm).
[The file location for Drupal's asset aggregation system is now configurable](https://www.drupal.org/node/3328126). It can be set in settings.php via $settings['file_assets_path']. Existing sites will continue to use the public files location.
A new setting $settings['sa_core_2023_004_phpinfo_flags'] in default.settings.php has been added to [configure the behaviour of admin/reports/status/php](https://www.drupal.org/node/3360166).

trackleft commented:
It would probably be worth checking block admin pages and workflow because of a number of these issues being resolved in 10.1

https://www.drupal.org/project/ideas/issues/3318110

**We have** https://github.com/az-digital/az_quickstart/blob/edb675a20efe483c7a6c0f3e166126f4dadc7e16/config/install/views.view.block_content.yml#L523
**Drupal core just added** 
https://git.drupalcode.org/project/drupal/-/blob/11.x/core/modules/block_content/config/optional/views.view.block_content.yml?ref_type=heads

joeparsons commented:
The messed up node edit form layout seems very reminiscent of an old issue we addressed in https://github.com/az-digital/az_quickstart/pull/1726 by adding a custom style library to `az_core` with style overrides to claro.

danahertzberg commented:
This is the culprit for the remaining claro width issue.

```
@media (min-width: 61rem)
.layout-region--node-main, .layout-region--node-footer {
    grid-column: 1;
    margin-inline: auto;
    width: min(52rem,100%);
}
```

`Width` is not recognized, but removing it fixes the problem.

trackleft commented:
Seeing this error: 
<img width="1697" alt="image" src="https://github.com/az-digital/az_quickstart/assets/1023167/90d5aa18-526a-4a07-8bc9-1401e014d484">


Transaction isolation level
READ-COMMITTED
For this to work correctly, all tables must have a primary key. The following table(s) do not have a primary key: honeypot_user. See the [setting MySQL transaction isolation level](https://www.drupal.org/docs/system-requirements/setting-the-mysql-transaction-isolation-level) page for [more information](/admin/config/system/cron).


Issue on drupal.org. https://www.drupal.org/project/honeypot/issues/2943526

danahertzberg commented on modules/custom/az_core/css/claro-node-form-overrides.css:
```diff
@@ -1,21 +0,0 @@
-  @see https://www.drupal.org/project/drupal/issues/3184667
-*/
-
-@media screen and (min-width: 61rem) {
```
```
@media (min-width: 61rem)
.layout-region--node-main, .layout-region--node-footer {
    margin-inline: 0;
    width: initial;
}
```

camikazegreen commented:
Minor issue in the editing sidebar that can probably be fixed with a followup:
<img width="361" alt="Screenshot 2023-09-20 at 10 11 52 AM" src="https://github.com/az-digital/az_quickstart/assets/1509771/edca0a8d-7625-4fde-b93f-c67679552c07">

joeparsons commented:
Worth noting that the patch added for the Honeypot primary key includes the changes committed to the dev version of that module for this other issue as well:
https://www.drupal.org/project/honeypot/issues/3121331

joeparsons commented:
> It would probably be worth checking block admin pages and workflow because of a number of these issues being resolved in 10.1
> 
> https://www.drupal.org/project/ideas/issues/3318110
> 
> **We have**
> 
> https://github.com/az-digital/az_quickstart/blob/edb675a20efe483c7a6c0f3e166126f4dadc7e16/config/install/views.view.block_content.yml#L523
> 
> 
> **Drupal core just added**
> https://git.drupalcode.org/project/drupal/-/blob/11.x/core/modules/block_content/config/optional/views.view.block_content.yml?ref_type=heads

Thanks @trackleft.  Yes, it looks like we had added the content blocks view as a tab on the main admin content view page in #645 but that seems to be redundant now that core is also providing effectively the same thing:

![Screenshot 2023-09-20 at 6 30 58 PM](https://github.com/az-digital/az_quickstart/assets/471936/31f3079a-4c48-4ff6-bc6e-90bafbe4bfe9)

I've updated this PR to remove the old view config added in #645 in favor of what now exists in core.

I also removed some claro block configs that don't appear to be necessary anymore similar to what was done in core's standard profile: https://www.drupal.org/i/3278565

joeparsons commented on .lando.yml:
```diff
@@ -104,7 +104,7 @@ tooling:
   eslint:
     service: node
-    cmd: yarn run eslint --color
+    cmd: yarn run eslint --color .
```
```suggestion
    cmd: yarn run eslint --color
```

joeparsons commented:
Follow-up task: re-export all configuration
